File: BestTimeToSellStock.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        
        if(prices == null || prices.Length == 0)
        {
            return 0;
        }
        
        int [] res = new int [prices.Length];
        int diff =  -prices[0];
        
        for(int i = 1; i < prices.Length; i++)
        {
            res[i] = Math.Max(res[i - 1], prices[i] + diff);
            diff = Math.Max(diff, -prices[i]);
        }
        return res[prices.Length - 1];
    }
    
}

File: BestTimeToSellStock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minP, maxP = float('inf'), 0
        
        for price in prices:
            maxPSell = price - minP
            maxP = max(maxP, maxPSell)
            minP = min(minP, price)
        return maxP
